interconnect/circuit.py

```diff
@@ -486,31 +499,75 @@ def __init__(self, tiles: Dict[int, Tile],
         # doesn't have any mux
+        self.__add_tile_id()
         self.__add_config()
         self.__add_stall(stall_signal_width)
         self.__add_reset()
 
+        # lift ports if there is empty sb
+        self.__lift_ports()
+
         # tile ID
         self.instance_name = f"Tile_X{self.x:02X}_Y{self.y:02X}"
 
+    def __add_tile_id(self):
+        self.add_port("tile_id",
+                      magma.In(magma.Bits(self.tile_id_width)))
+
     def __add_stall(self, stall_signal_width: int):
-        # automatically add stall signal and connect it to the core if the
-        # core supports it
+        # automatically add stall signal and connect it to the features if the
+        # feature supports it
         self.add_ports(stall=magma.In(magma.Bits(stall_signal_width)))
-        if "stall" in self.core.ports.keys():
-            self.wire(self.ports.stall, self.core.ports.stall)
+        stall_ports = []
+        for feature in self.features():
+            if "stall" in feature.ports.keys():
+                stall_ports.append(feature.ports.stall)
+        # some core may not expose the port as features, such as mem cores
+        if self.core is not None and "stall" in self.core.ports and \
+                self.core.ports.stall not in stall_ports:
+            stall_ports.append(self.core.ports.stall)
+        for stall_port in stall_ports:
+            self.wire(self.ports.stall, stall_port)
 
     def __add_reset(self):
-        if "reset" in self.core.ports.keys():
-            self.wire(self.ports.reset, self.core.ports.reset)
+        # automatically add reset signal and connect it to the features if the
+        # feature supports it
+        self.add_ports(reset=magma.In(magma.AsyncReset))
+        reset_ports = []
+        for feature in self.features():
+            if "reset" in feature.ports.keys():
+                reset_ports.append(feature.ports.reset)
+        # some core may not expose the port as features, such as mem cores
+        if self.core is not None and "reset" in self.core.ports and \
+                self.core.ports.reset not in reset_ports:
+            reset_ports.append(self.core.ports.reset)
+
+        for reset_port in reset_ports:
+            self.wire(self.ports.reset, reset_port)
+
+    def __should_add_config(self):
+        # a introspection on itself to determine whether to add config
+        # or not
+        for feature in self.features():
+            if "config" in feature.ports:
+                return True
+            else:
+                # if the feature doesn't have config port, it shouldn't have
+                # reset either, although the other way around may be true
+                # that is, a feature may have some internal states that need
+                # to reset, but not necessarily has config port
+                assert "reset" not in feature.ports
+        return False
 
     def __add_config(self):
```

Comment:
Even if `__should_add_config` is `True`, you cannot iterate all features (line 582) and wire `config_data` to all features, because not all features in the tile has configuration.
For example, if io_tile has connection box because it is connected to  multiple tracks from `switchbox` in the neighbor tile (so it needs configuration), io_core does not have configuration, so line 582 will cause an error.